GetPot Unit Tests + GetPot::have_section()

pbauman opened the pull request:
Added CppUnit tests for GetPot. Actual source file has to be generated by configure since CppUnit doesn't provide access to argv (annoying) and GetPot only takes file names in the constructor, no streams (so couldn't do stringstream). Motivation for the unit tests was adding a new API to GetPot - have_section(). It's useful to be able to test whether a given section is in the input file. To facilitate a binary search instead of a linear search, I added a sort() to the section_list at the end of parsing.

moosebuild commented:
Results of testing 9cda74d49b11195feb230bff56ecd64fadb9b08f using libmesh_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12425

moosebuild commented:
Results of testing 9cda74d49b11195feb230bff56ecd64fadb9b08f using libmesh_PR_test_dbg recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12426

jwpeterson commented:
FYI, looks like the failure was in the new unit test...
```

1) test: GetPotTest::testSections (F) line: 73 base/getpot_test.C
assertion failed
- Expression: input.have_section("Section2/Subsection2/Subsection3")
```

pbauman commented:
Thanks. Getting back from dinner and am going to take a look. Might've forgotten a last commit in my hurry to get out.

> On Feb 27, 2015, at 6:36 PM, "John W. Peterson" <notifications@github.com> wrote:
> 
> FYI, looks like the failure was in the new unit test...
> 
> 
> 1) test: GetPotTest::testSections (F) line: 73 base/getpot_test.C
> assertion failed
> - Expression: input.have_section("Section2/Subsection2/Subsection3")
> —
> Reply to this email directly or view it on GitHub.
>

pbauman commented:
OK, weird. I could've sworn those were passing before I left the office. The network is conveniently down so I can't see what happened. Nevertheless, fixed and squashed that commit and added documentation that when querying the section, it *must* end in a "/". e.g. sections names are "/Section1/Section2/", etc.

pbauman commented:
Would it be better to have the method look at your string and append the "/" if it's missing? I was trying to play it safe by not mucking with what was coming in, but I'm not thinking of any cases that it could bite you. Could save a lot of headaches...

moosebuild commented:
Results of testing 264b3c367f0a02f9fdbe178348d7dea2dc419710 using libmesh_PR_test recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12455

pbauman commented:
Wat?!

I don't see how I'd be hosing y'all. Looked like a few Exodus diff problems. Are these new functions somehow colliding? Perhaps something because I bootstrapped? Other?

moosebuild commented:
Results of testing 264b3c367f0a02f9fdbe178348d7dea2dc419710 using libmesh_PR_test_dbg recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12456

jwpeterson commented:
Hmm... looks like you are now sorting the entries in the GetPot object?  I'm not sure, but I think `materials/derivativematerialinterface.property_names` is a new test that relies on input file ordering.  So that might be the issue...

pbauman commented:
Yeah, I sorted the section_list, but at parsing time so that it was sorted before anyone could call 'GetPot::get_section_names()'. OK, I'll revert the sorting and use a linear search in have_section().

moosebuild commented:
Results of testing 250ffcd075ca0bc85af91400bf0945b4feaac9ed using libmesh_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12457

moosebuild commented:
Results of testing 250ffcd075ca0bc85af91400bf0945b4feaac9ed using libmesh_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12458

roystgnr commented on include/base/getpot.h:
```diff
@@ -202,6 +202,24 @@ class GetPot
   /**
+   * Check for a section name. When querying, the section_name
+   * MUST end in a forward slash. e.g.
+   * /Section1/
```
Kind of fragile, no?  Can we test for a trailing forward slash and add it internally if absent?

roystgnr commented:
That's my only nitpick.  :+1:

moosebuild commented:
Results of testing a94de7804bd9a17fdf4efc904d30a859e873ed44 using libmesh_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12459

moosebuild commented:
Results of testing a94de7804bd9a17fdf4efc904d30a859e873ed44 using libmesh_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12460

pbauman commented:
Addressed @roystgnr's concern and MooseBuild is passing. If a different strategy for handling out-of-source builds + CppUnit's unwillingness to give you argc/argv, I'll address @jwpeterson's concern. Merging.
